server/src/test/java/org/elasticsearch/index/analysis/NamedAnalyzerTests.java

```diff
@@ -0,0 +1,79 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.index.analysis;
+
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.TokenStream;
+import org.elasticsearch.index.mapper.MapperException;
+import org.elasticsearch.test.ESTestCase;
+
+public class NamedAnalyzerTests extends ESTestCase {
+
+    public void testCheckAllowedInMode() {
+        try (NamedAnalyzer testAnalyzer = new NamedAnalyzer("my_analyzer", AnalyzerScope.INDEX,
+                createAnalyzerWithMode("my_analyzer", AnalysisMode.INDEX_TIME), Integer.MIN_VALUE)) {
+            testAnalyzer.checkAllowedInMode(AnalysisMode.INDEX_TIME);
+            MapperException ex = expectThrows(MapperException.class, () -> testAnalyzer.checkAllowedInMode(AnalysisMode.SEARCH_TIME));
+            assertEquals("analyzer [my_analyzer] contains filters [my_analyzer] that are not allowed to run in search time mode.",
+                    ex.getMessage());
+            ex = expectThrows(MapperException.class, () -> testAnalyzer.checkAllowedInMode(AnalysisMode.ALL));
+            assertEquals("analyzer [my_analyzer] contains filters [my_analyzer] that are not allowed to run in all mode.", ex.getMessage());
+        }
+
+        try (NamedAnalyzer testAnalyzer = new NamedAnalyzer("my_analyzer", AnalyzerScope.INDEX,
+                createAnalyzerWithMode("my_analyzer", AnalysisMode.SEARCH_TIME), Integer.MIN_VALUE)) {
+            testAnalyzer.checkAllowedInMode(AnalysisMode.SEARCH_TIME);
+            MapperException ex = expectThrows(MapperException.class, () -> testAnalyzer.checkAllowedInMode(AnalysisMode.INDEX_TIME));
+            assertEquals("analyzer [my_analyzer] contains filters [my_analyzer] that are not allowed to run in index time mode.",
+                    ex.getMessage());
+            ex = expectThrows(MapperException.class, () -> testAnalyzer.checkAllowedInMode(AnalysisMode.ALL));
+            assertEquals("analyzer [my_analyzer] contains filters [my_analyzer] that are not allowed to run in all mode.", ex.getMessage());
+        }
+
+        try (NamedAnalyzer testAnalyzer = new NamedAnalyzer("my_analyzer", AnalyzerScope.INDEX,
+                createAnalyzerWithMode("my_analyzer", AnalysisMode.ALL), Integer.MIN_VALUE)) {
+            testAnalyzer.checkAllowedInMode(AnalysisMode.ALL);
+            testAnalyzer.checkAllowedInMode(AnalysisMode.INDEX_TIME);
+            testAnalyzer.checkAllowedInMode(AnalysisMode.SEARCH_TIME);
+        }
+    }
+
+    private Analyzer createAnalyzerWithMode(String name, AnalysisMode mode) {
```

Comment:
It's great that we can have a unit test for this but I wonder if an integ test with a analyzer plugin should also be tested ? This way you could test the high level behavior when you mix a custom token filter with specific analysis mode and other token filters that use the ALL mode ?

Reply:
Sorry, I don't really see the benefit of an integration test at this point. What would an analyzer plugin add that is not tested here?
As soon as we use different modes, e.g. like I was already doing previously in this PR with the SynonymTokenFilter, it will be tested in integration and rest tests, but I don't see any benefit adding it with other mocked filters here.

Reply:
> What would an analyzer plugin add that is not tested here?

I think it would provide a good example of how to use and a way to check if the mode is correctly propagated when you mix multiple analyzer. I agree that most of this is already tested in the mock test so I don't have strong feeling about it either.